aten/src/ATen/native/cuda/Activation.cu

```diff
@@ -2,9 +2,239 @@
 #include "ATen/NativeFunctions.h"
 #include "ATen/Dispatch.h"
 #include "ATen/cuda/CUDAApplyUtils.cuh"
+#include "ATen/cuda/detail/IndexUtils.cuh"
+
 
 namespace at { namespace native {
 
+// -----------------------------------
+// prelu forward
+// -----------------------------------
+template <typename scalar_t>
+void prelu_cuda_kernel_share_weights(
+  const Tensor& input,
+  Tensor& result,
+  const scalar_t* weight_data) {
+
+  at::cuda::CUDA_tensor_apply2<scalar_t, scalar_t>(
+    input,
+    result,
+    [=] __device__ (
+      const scalar_t& input_val,
+      scalar_t& result_val) {
+        result_val = (input_val > 0) ? input_val : *weight_data * input_val;
+  });
+}
+
+template <typename scalar_t>
+__global__ void prelu_cuda_kernel_multi_weights(
+  scalar_t* result_data,
+  const scalar_t* input_data,
+  const scalar_t* weight_data,
+  int64_t input_stride0,
+  int64_t input_stride1,
+  int64_t input_numel) {
+
+  int64_t linearId = blockIdx.x * blockDim.x + threadIdx.x;
+  if (linearId >= input_numel) return;
+
+  // multiply values at each channel with weight[channel_index]
+  int64_t channel = (linearId % input_stride0) / input_stride1;
+  scalar_t input_data_val = input_data[linearId];
+  result_data[linearId] = (input_data_val > 0) ? input_data_val : weight_data[channel] * input_data_val;
+}
+
+Tensor prelu_cuda(const Tensor& self, const Tensor& weight_) {
+  AT_CHECK(self.is_cuda());
+  AT_CHECK(weight_.is_cuda());
+
+  auto input = self.contiguous();
+  auto weight = weight_.contiguous();
+
+  AT_CHECK(input.is_contiguous());
+  AT_CHECK(weight.is_contiguous());
+
+  int64_t weight_num = weight.numel();
+  Tensor result = at::empty_like(input);
+  auto strides = input.strides();
+
+  // case1: shared weight for all channels
+  if (weight_num == 1) {
```

Comment:
CUDA kernel for weight.numel() == 1 is still slower than previous implementation, not sure why... can it be the difference between `THC_pointwiseApply2` and `CUDA_tensor_apply2`?

https://github.com/pytorch/pytorch/blob/b91b15d86e274611f448caa3f9b73eb4bf24a625/aten/src/THCUNN/generic/PReLU.cu#L17-L20

https://github.com/pytorch/pytorch/blob/b91b15d86e274611f448caa3f9b73eb4bf24a625/aten/src/THCUNN/PReLU.cu#L9-L23